pipeline-runner/R/qc-4-filter_gene_umi_outlier.R

```diff
@@ -25,85 +25,80 @@
-      outliers <- rownames(df)[df$genes > pb$upr | df$genes < pb$lwr]
-
-      # Keep the ones that are not outlier
-      # Because we have subseted the object before creating the dataframes, and just selected the outlier values from that df
-      # In the end we wil only remove values that are in the desired sample.
-      scdata.filtered <- subset_safe(scdata, colnames(scdata)[!colnames(scdata) %in% outliers])
-      # Similarly, when creating the plot from the df, we will just be using the values for the required sample.
-      plot1_data <- unname(purrr::map2(df$molecules, df$genes, function(x, y) {
-        c("log_molecules" = x, "log_genes" = y)
-      }))
-      plot1_data <- purrr::map2(plot1_data, unname(pb$lwr), function(x, y) {
-        append(x, c("lower_cutoff" = y))
-      })
-      plot1_data <- purrr::map2(plot1_data, unname(pb$upr), function(x, y) {
-        append(x, c("upper_cutoff" = y))
-      })
-      # have downsampling done
-
-      # Downsample plotData
-      num_cells_to_downsample <- downsample_plotdata(ncol(sample_subset), num_cells_to_downsample)
-
-      set.seed(123)
-      cells_position_to_keep <- sample(1:ncol(sample_subset), num_cells_to_downsample, replace = FALSE)
-      cells_position_to_keep <- sort(cells_position_to_keep)
-      plot1_data <- plot1_data[cells_position_to_keep]
-    }
-  } else {
-    scdata.filtered <- scdata
-    plot1_data <- list()
-  }
+  type <- config$filterSettings$regressionType
 
-  config$filterSettings$regressionTypeSettings[[config$filterSettings$regressionType]][["p.level"]] <- p.level
+  # get p.level and update in config
+  # defaults from "gene.vs.molecule.cell.filter" in pagoda2
+  if (!is.null(config$auto) && config$auto)
+    p.level <- min(0.001, 1 / ncol(scdata))
+  else
+    p.level <- config$filterSettings$regressionTypeSettings[[type]]$p.level
 
-  # Scatter plot which is composed of:
-  # x-axis: log_10_UMIs
-  # y-axis: log_10_genes
-  # bands that are conformed with the upper_cutoff and the lower_cutoff.
-  guidata <- list()
-  guidata[[generate_gui_uuid(sample_id, task_name, 0)]] <- plot1_data
+  config$filterSettings$regressionTypeSettings[[type]]$p.level <- p.level
+
+  # defaults if not enabled
+  scdata.filtered <- scdata
+  plot_data <- list()
+
+  # barcodes for this sample
+  sample_barcodes <- colnames(scdata)[scdata$samples == sample_id]
+
+  if (config$enabled && length(sample_barcodes)) {
+
+    # subset this sample
+    sample_subset <- scdata[, sample_barcodes]
+
+    # regress log10 molecules vs genes
+    data <- data.frame(
+      log_molecules = log10(sample_subset$nCount_RNA),
+      log_genes = log10(sample_subset$nFeature_RNA),
+      row.names = colnames(sample_subset)
+    )
+
+    data <- data[order(data$log_molecules), ]
+    fit <- lm(log_genes ~ splines::bs(log_molecules), data = data)
+
+    # get the interval based on p.level parameter
+    preds <- suppressWarnings(predict(fit, interval = "prediction", level = 1 - p.level))
```

Comment:
Why suppresing warnings?

Reply:
same as before. `predict` complains if you don't supply `newdata`